### PR TITLE
KAFKA-17613: Remove ZK migration code

### PR DESCRIPTION
**Pull request by cmccabe**
Remove the controller machinery for doing ZK migration in Kafka 4.0.

As a note, I've taken the approach of leaving KafkaConfig.migrationEnabled but hard-coding it to
false here. That was because I didn't want to mess around with KafkaServer.scala and related classes.
Since they're getting removed soon anyway, it would be a waste.

### CODE REVIEW

**Review by mumrah — COMMENTED** [review]
I don't see anything obvious, but this might collide with https://github.com/apache/kafka/pull/17284

**Comment by cmccabe** [comment]
As a note, I've taken the approach of leaving `KafkaConfig.migrationEnabled` but hard-coding it to `false` here. That was because I didn't want to mess around with `KafkaServer.scala` and related classes. Since they're getting removed soon anyway, it would be a waste.

**Comment by mumrah** [comment]
There are some test failures. I think they'll be fixed after we merge #17284

**Comment by mumrah** [comment]
@cmccabe the other removal PR is merged now, can you merge in trunk?

File: metadata/src/main/java/org/apache/kafka/controller/metrics/QuorumControllerMetrics.java [hunk 1]
```diff
@@ -76,8 +70,6 @@ public class QuorumControllerMetrics implements AutoCloseable {
     private final AtomicLong dualWriteOffset = new AtomicLong(0);
```
Comment (chia7712) [inline metadata/src/main/java/org/apache/kafka/controller/metrics/QuorumControllerMetrics.java]: Is dual-write metrics still used in production code after `KRaftMigrationDriver` is removed?

File: core/src/main/scala/kafka/server/ApiVersionManager.scala [hunk 2]
```diff
@@ -112,7 +107,7 @@ class SimpleApiVersionManager(
       setSupportedFeatures(brokerFeatures).
       setFinalizedFeatures(currentFeatures.finalizedFeatures()).
       setFinalizedFeaturesEpoch(currentFeatures.finalizedFeaturesEpoch()).
-      setZkMigrationEnabled(zkMigrationEnabled).
+      setZkMigrationEnabled(false).
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/ApiVersionManager.scala:110]: why not removing the `setZkMigrationEnabled`?

**Review by chia7712 — COMMENTED** [review]
@cmccabe thanks for this cleanup.

File: metadata/src/main/java/org/apache/kafka/controller/metrics/QuorumControllerMetrics.java (hunk 1 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/controller/metrics/QuorumControllerMetrics.java]: good point, I will remove this.

File: core/src/main/scala/kafka/server/ApiVersionManager.scala (hunk 2 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/ApiVersionManager.scala:110]: We cannot remove `zkMigrationEnabled` from the RPC without breaking backwards compatibility. Although it does default to `false`, so we could leave out this line (which I think might be more confusing to read, but would be technically correct).

Comment (chia7712) [reply core/src/main/scala/kafka/server/ApiVersionManager.scala:110]: Maybe we can hardcode `zkMigrationReady` to `false` when building `ApiVersionsResponse` (`ApiVersionsResponse#Builder#build`) and then remove this setter ?

Comment (cmccabe) [reply core/src/main/scala/kafka/server/ApiVersionManager.scala:110]: Wouldn't that make testing harder? It seems fine to just set it here.

**Comment by cmccabe** [comment]
I've updated this to remove the lingering reference to dual write offset in the metrics class.

[edit: also merged in trunk as @mumrah suggested. Thanks for the reviews, everyone.]

**Review by mumrah — APPROVED** [review]
LGTM

**Comment by chia7712** [comment]
I believe those failing tests can be removed if they are specifically designed to run with ZooKeeper migration

**Comment by chia7712** [comment]
@cmccabe Could you please fix the conflicts? the failed tests were removed by #17318

**Pull request merged**